pkg/controllers/operationjob/restart/restart.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2024 The KusionStack Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restart
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	appsv1alpha1 "kusionstack.io/operating/apis/apps/v1alpha1"
+	. "kusionstack.io/operating/pkg/controllers/operationjob/opscontrol"
+	ojutils "kusionstack.io/operating/pkg/controllers/operationjob/utils"
+	"kusionstack.io/operating/pkg/controllers/utils/podopslifecycle"
+)
+
+type ContainerRestartControl struct {
+	*OperateInfo
+}
+
+func (p *ContainerRestartControl) ListTargets() ([]*OpsCandidate, error) {
+	var candidates []*OpsCandidate
+	podOpsStatusMap := ojutils.MapOpsStatusByPod(p.OperationJob)
+	for _, target := range p.OperationJob.Spec.Targets {
+		var candidate OpsCandidate
+		var pod corev1.Pod
+
+		// fulfil pod
+		candidate.PodName = target.PodName
+		err := p.Client.Get(p.Context, types.NamespacedName{Namespace: p.OperationJob.Namespace, Name: target.PodName}, &pod)
+		if err == nil {
+			candidate.Pod = &pod
+		} else if errors.IsNotFound(err) {
+			candidate.Pod = nil
+		} else {
+			return candidates, err
+		}
+
+		// fulfil containers
+		candidate.Containers = target.Containers
+		if len(target.Containers) == 0 && candidate.Pod != nil {
+			// restart all containers
+			var containers []string
+			for _, container := range candidate.Pod.Spec.Containers {
+				containers = append(containers, container.Name)
+			}
+			candidate.Containers = containers
+		}
+
+		// fulfil or initialize opsStatus
+		if opsStatus, exist := podOpsStatusMap[target.PodName]; exist {
+			candidate.PodOpsStatus = opsStatus
+		} else {
+			candidate.PodOpsStatus = &appsv1alpha1.PodOpsStatus{
+				PodName:  target.PodName,
+				Progress: appsv1alpha1.OperationProgressPending,
+			}
+		}
+
+		candidates = append(candidates, &candidate)
+	}
+	return candidates, nil
+}
+
+func (p *ContainerRestartControl) OperateTarget(candidate *OpsCandidate) error {
+	// mark candidate ops started is not started
+	if IsCandidateOpsPending(candidate) {
+		candidate.PodOpsStatus.Progress = appsv1alpha1.OperationProgressProcessing
+	}
+
+	// skip if candidate ops finished, or pod and containers do not exist
+	_, containerNotFound := ojutils.ContainersNotFoundInPod(candidate.Pod, candidate.Containers)
```

Comment:
`candidate.Pod == nil` has been checked at below line. This line should be replaced in following IF statement.